### PR TITLE
OOP school library: set up associations

### PR DESCRIPTION
**Pull request by CindyDorantes**
## Project requirements   :dart: :dart:
- Create a class `Classroom` with the following:
  - `@label` instance variable, should be initialized in the constructor.
  - Setter and getter for `@label` (remember about `attr_accessor`).
- Create the `has-many`/`belongs-to` relationship between `Classroom` and `Student`. The following should be implemented:
  - Create the `has-many` side (a classroom has many students).
  - Create the `belongs-to` side (a student belongs to a classroom).
  - Make sure that when adding a student to a classroom it also sets the classroom for the student.
  - Make sure that when setting the classroom for a student it also adds it to the classrooms' students.
- Create a class `Book` with the following:
  - `@title` and `@author` instance variables, should be initialized in the constructor.
  - Setters and getters for instance variables (remember about `attr_accessor`).
- Create a class `Rental` with the following:
  - `@date` instance variable, should be initialized in the constructor.
  - Setter and getter for `@date` (remember about `attr_accessor`).
- Create the `many-to-many` (also `has-many-through`) relationship between `Person` and `Book` using the intermediate class `Rental`. The following should be implemented:
  - Create the `has-many` side of `Book` and `Rental` (a book has many rentals).
  - Create the `belongs-to` side of `Rental` and `Book` (a rental belongs to a book).
  - Create the `has-many` side of `Person` and `Rental` (a person has many rentals).
  - Create the `belongs-to` side of `Rental` and `Person` (a rental belongs to a person).
  - Modify the constructor of `Rental` so `Book` and `Person` are set in it.

### CODE REVIEW

File: book.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
+end
```
Comment (Nemwel-Boniface) [inline book.rb]: - [x] It seems like you might have forgotten to implement the "many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental"  the a book has many rentals association.

image

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
```
Comment (Nemwel-Boniface) [inline book.rb]: - [x] This should be within the "attr_accessor" and not the "attr_reader" thus making it something like this : ```attr_accessor :title, :author, :rentals```

File: rental.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Rental
+  attr_accessor :date
+  attr_reader :book, :person
```
Comment (Nemwel-Boniface) [inline rental.rb]: - [x] This should be within the "attr_accessor" and not the "attr_reader" thus making it something like this : ```attr_accessor :date, :book, :person```

File: rental.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+  def initialize(date, book, person)
+    @date = date
+    @book = book
+    book.rentals << self
+    @person = person
+    person.rentals << self
+  end
+end
```
Comment (Nemwel-Boniface) [inline rental.rb]: - [x] It seems like you forgot to implement the "A rental belongs to a book and also belongs to a person" relationship. Kindly add it.

image

File: person.rb [hunk 5]
```diff
@@ -1,9 +1,9 @@
-require_relative 'nameable'
-require_relative 'capitalize_decorator'
-require_relative 'trimmer_decorator'
+require_relative './nameable'
+require_relative './capitalize_decorator'
+require_relative './trimmer_decorator'
 
 class Person < Nameable
-  attr_reader :id
+  attr_reader :id, :rentals
```
Comment (Nemwel-Boniface) [inline person.rb]: - [x] This should be within the "attr_accessor" and not the "attr_reader" thus making it something like this : ```attr_accessor :name, :age, :rentals``

File: person.rb [hunk 6]
```diff
@@ -1,9 +1,9 @@
-require_relative 'nameable'
-require_relative 'capitalize_decorator'
-require_relative 'trimmer_decorator'
+require_relative './nameable'
+require_relative './capitalize_decorator'
+require_relative './trimmer_decorator'
 
 class Person < Nameable
```
Comment (Nemwel-Boniface) [inline person.rb:5]: - [ ] It seems you forgot to implement the "A person has many rentals" relationship. Kindly include it.

image

**Review by Nemwel-Boniface — CHANGES_REQUESTED** [review]
## Hello @CindyDorantes      :wave: , 
I'm Nemwel your reviewer for this PR :man_beard: 

Good job working on your project so far! :clap: 
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights

Great job setting up the correct linters ✔️ 
Professional documentation ✔️ 
Correct workflow ✔️ 
Meaningful commit messages ✔️  


## ♻️ Required Changes ♻️

_Check the comments under the review._

## [Optional] suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding! :partying_face: 👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me **@Nemwel-Boniface** in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: book.rb (hunk 1 above)
Comment (CindyDorantes) [reply book.rb]: Hi @Nemwel-Boniface 
I indeed implemented the many-to-many relationship using the intermediate class Rental as requested by the project requirements. So when a new Rental instance is created, the initialization takes action to assign the rental to the person and the book declared. You can try running the following code:
```
require_relative './person'
require_relative './book'
require_relative './rental'

andy = Person.new(33, 'Andy')
cindy = Person.new(33, 'Cindy')

book1 = Book.new('The Iliad', 'Homer')
book2 = Book.new('The Complete Stories', 'Isaac Asimov')

rental01 = Rental.new('2022-10-12', book1, andy)
rental02 = Rental.new('2022-10-12', book2, andy)
rental03 = Rental.new('2022-08-10', book1, cindy)
rental04 = Rental.new('2022-08-12', book2, cindy)

puts andy.rentals.map { |rental| rental.book.title }
puts cindy.rentals.map { |rental| rental.book.title }
puts book1.rentals.map { |rental| rental.person.name }
puts book1.rentals.map { |rental| rental.person.name }
```
There you will see that it is possible to have "many-to-many" relationship between Person and Book

File: book.rb (hunk 2 above)
Comment (CindyDorantes) [reply book.rb]: Done! Thanks

File: rental.rb (hunk 3 above)
Comment (CindyDorantes) [reply rental.rb]: Done! Thanks

File: rental.rb (hunk 4 above)
Comment (CindyDorantes) [reply rental.rb]: Hi @Nemwel-Boniface 
Please refer to my comment on the first change requested in this PR

File: person.rb (hunk 5 above)
Comment (CindyDorantes) [reply person.rb]: Done! Thanks

File: person.rb (hunk 6 above)
Comment (CindyDorantes) [reply person.rb:5]: Hi @Nemwel-Boniface 
Please refer to my comment on the first change requested in this PR

**Review by victorokoroji — APPROVED** [review]
Hi @CindyDorantes, 

# ✔️   Status: APPROVED 🏆 🎉 🎊 

You have done a great job on this project, keep it up 👍 🥇. Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
congratulations!  🎉💯🌟 

## Top highlight 🎯

- The Pull Request is following the GitHub Flow ✅ 
- The PR title and description are professional and descriptive ✅ 
- The commit messages are meaningful ✅ 



## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**